File: TopInterview150/Intervals/MergeIntervals/solution.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> output;
        sort(intervals.begin(), intervals.end(), [](const vector<int>& a, const vector<int>& b) {
            return a[0] < b[0];
        });

        int index = 0;
        for (int i = 0; i < intervals.size(); i++)//for each vector
        {
            if (index == 0 || output[index-1][1] < intervals[i][0])//if the next start is greater than the current end (no overlap)
            {
                //add to output
                output.push_back(intervals[i]);
                index++; 
            }
            else 
            {//update the previous index interval's end if appropriate
                output[index-1][1] = max(output[index-1][1], intervals[i][1]);
            }
        }


        return output;
    }
};

File: TopInterview150/Intervals/MergeIntervals/solution.py
class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: List[List[int]]
        """
        output = []
        intervals.sort(key = lambda x: x[0])
        print(intervals)
        
        index = 0
        for i in range(len(intervals)):
            """ if the first index or no overlap"""
            if index == 0 or output[index-1][1] < intervals[i][0]:
                output.append(intervals[i])
                index+=1
            else:
                output[index-1][1] = max(output[index-1][1], intervals[i][1])

        
        return output